tests/python_compiler/transforms/quantum/test_xdsl_parity_synth.py

```diff
@@ -0,0 +1,490 @@
+        # Sample random CNOTs (by index into above list) and rotation angles
+        cnots = [
+            np.random.choice(len(all_cnots), size=n, replace=True) for _ in range(num_parities)
+        ]
+        thetas = np.random.random(num_parities)
+        # Make PL circuit
+        circuit = sum(
+            [
+                [all_cnots[i] for i in sub_circuit] + [qml.RZ(x, j % n)]
+                for j, (sub_circuit, x) in enumerate(zip(cnots, thetas, strict=True))
+            ],
+            start=[],
+        )
+        # Compute IR
+        _, P, angles = phase_polynomial(qml.tape.QuantumScript(circuit), wire_order=range(n))
+
+        angles_ = list(angles)
+        # Synthesize parity network and compute new PL circuit from it
+        new_circuit, inv_parity_matrix = _parity_network_synth(P)
+        new_circuit = sum(
+            [
+                [qml.CNOT(_cnot) for _cnot in sub_circuit]
+                + [qml.RZ(angles_.pop(angle_idx), qubit_idx)]
+                for angle_idx, qubit_idx, sub_circuit in new_circuit
+            ],
+            start=[],
+        )
+        # Compute IR of new PL circuit
+        new_parity_matrix, new_P, new_angles = phase_polynomial(
+            qml.tape.QuantumScript(new_circuit), wire_order=range(n)
+        )
+        # Compare phase parities and make sure that the inv_parity_matrix is valid
+        assert_allclose(new_P @ new_angles, P @ angles)
+        assert_binary_matrix(inv_parity_matrix)
+        assert_equal((new_parity_matrix @ inv_parity_matrix) % 2, np.eye(n, dtype=int))
+
+
+def translate_program_to_xdsl(program):
+    """Translate an almost-xDSL-program into an xDSL program by replacing some shorthand notations."""
+    new_lines = []
+    for line in program.split("\n"):
+        if "INIT_QUBIT" in line:
+            i = int(line.strip().split(" ")[0][1:])
+            new_lines.extend(
+                [
+                    f'// CHECK: [[q{i}:%.+]] = "test.op"() : () -> !quantum.bit',
+                    f'%{i} = "test.op"() : () -> !quantum.bit',
+                ]
+            )
+        elif "_CNOT" in line:
+            bits = line.strip().split(" ")
+            new_bits = (
+                bits[:3] + ['quantum.custom "CNOT"()'] + bits[4:] + [": !quantum.bit, !quantum.bit"]
+            )
+            new_lines.append(" ".join(new_bits))
+        else:
+            new_lines.append(line)
+    return "\n".join(new_lines)
+
+
+class TestParitySynthPass:
+    """Unit tests for ParitySynthPass."""
+
+    pipeline = (ParitySynthPass(),)
+
+    def test_no_phase_polynomial_ops(self, run_filecheck):
+        """Test that nothing changes when there are no phase polynomial gates."""
+        program = """
+            func.func @test_func(%arg0: f64) {
+                %0 = INIT_QUBIT
+                // CHECK: [[q1:%.+]] = quantum.custom "Hadamard"() [[q0]] : !quantum.bit
+                // CHECK: quantum.custom "RX"(%arg0) [[q1]] : !quantum.bit
```

Comment:
why do we check anything about Hadamards? I might be confused about what the CHECK directive for `run_filecheck` is doing. Are we checking anything about Hadamards? Or just telling `run_filecheck` to check this line of the program, which just happens to initialize a Hadamard gate?

Reply:
The transform should not touch anything if there are no phase-polynomial ops. Therefore, we just CHECK that the same ops are there as we put into the input program (the two lines below with `%1 = ` and `%2 = `)

Reply:
I'll ask you later about the CHECK stuff, I think I'm still not sure what this does / how this works even with the cookbook description